### PR TITLE
Improve CTE materialization execution tests

### PR DESCRIPTION
**Pull request by misterjpapa**
Add method verifyResults and checkCTEInfoMatch to test whether CTEs are materialized by checking if the Explain plan has the expected CTEInfo element.

Existing tests were only asserting if results of executing the query are the same with and without using a CTE materialization strategy. These new methods go further and also check that the expected CTEs were materialized.

Refactor existing tests to use the verifyResults method or in the case of more complex queries the method checkCTEInfoMatch. In order for Explain plan to include the CTEInfo element, the system property VERBOSE_OPTIMIZER_INFO_ENABLED needs to be set for the session.

## Description
The existing CTE execution tests are only checking that the query results are the same with and without using a CTE materialization strategy during execution. Two new test helper methods verifyResults and checkCTEInfoMatch were introduced to also check the the Explain plan for the query execution included a CTEInfo element to verify that the CTEs are materialized.  

## Motivation and Context
Fix an open issue about improving existing CTE materialization tests to check beyond just the results of the query and verify the CTE materialization. See https://github.com/prestodb/presto/issues/21688

## Impact
The only changes are to improve existing tests. There is no public facing functionality impact.

## Test Plan
The only changes are to improve existing tests.

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by misterjpapa** [comment]
Looking at the ci/circleci linux-presto-e2e-tests errors.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:76]: can we have less abbreviated variable names?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:71]: I know we use this Language annotation in various places, but honestly I'm not a big fan of it. If I had my druthers, I'd drop it.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1275]: move this to a constant to avoid recomputing

**Comment by jaystarshot** [comment]
Thanks for your contribution! 
> Refactor the rest of the tests (45 remaining tests)
Wondering if there was any blocker for these, if not can we include them in the same commit?

**Comment by misterjpapa** [comment]
No blocker Jay. I didn't want to change all 55 tests without getting confirmation from the community that I was heading in the right direction. I will address elharo's comments above and refactor the rest of the tests in the same commit.  With regards to elharo's comment above about "Language annotation in various places" for consistently sake should I just revert the annotation additions for the 10 tests so far or continue to add the annotation to the rest of the test methods? I am also not a big fan of the Language annotation, but it appeared to be part of the convention/warning for this codebase so I decided to follow it.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (misterjpapa) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java]: I've moved the CTEInfoValue objects creation inline with the ImmutableList similar to the paradigm in other tests which circumvents the need to define variable names.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (misterjpapa) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java]: I've removed the Language annotation as it wasn't originally used in this class. If someone objects my suggestion would be to open a new issue so we can decide then.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (misterjpapa) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java]: Good idea - since latest changes.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (elharo) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:77]: doesn't matter a great deal but I tend to prefer Arrays.asList to ImmutableList.of to minimize thrid party dependencies

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (jaystarshot) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java]: https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md
our developer guide does seem to prefer immutable list

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (jaystarshot) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1270]: Can you please add a materialized boolean here? Just in case we might want to assert non materialized ctes

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (jaystarshot) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1267]: Also can we just reuse the CTEInformation class from the session?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (elharo) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:77]: "Prefer Immutable collections in Guava when possible." My experience is that it's not possible or advisable as much as most people think. I've had too many really ugly debugging sessions in production systems because an immutable type snuck into some code that expected to be able to mutate it. The root problem is that the Java Collections API was misdesigned 25 years ago in a way that violates the Liskov substitution principle, but we can't fix that here so the best we can do is not use types that don't implement the full List API. If we do want to use immutable types, we should make sure the API reflects that so folks aren't surprised when it doesn't mutate, but then we tightly couple to a third party API, which we also don't want to do. There's no good answer here.

All that being said, it's also possible to use Collections.unmodifiableList(Arrays.asList(...)) and retain immutability while avoiding third party dependencies. In this case since it's test code I'd prefer to keep it smaller as the immutability really isn't needed, but up to you.

Comment (jaystarshot) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:77]: Sg, maybe we should update the guide

Comment (misterjpapa) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:77]: +1 on "really ugly debugging sessions in production systems because an immutable type snuck into some code that expected to be able to mutate it", but as Jay mentioned it's explicitly called out in the Developer Guide that Immutable types should be used. Perhaps suggest an update to the Developer Guide for more nuisance use of the Immutable collections if we can come to a consensus.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (misterjpapa) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1267]: Good point. It didn't register in my mind to use the CTEInformation class. I can created a CTEInformation rather than a CTEInfoValue, but I won't have access to a cteId value so it will stay null or match the name. Is that acceptable - otherwise I will just continue to use the CTEInfoValue inner class and extend it with the isMaterialized boolean? As mentioned in the earlier comment, it would have been great to just use the CTEInformation from the Session object directly, but that Session is created in-line when the DispatchManager is creating an internal query which the tests doesn't have access to.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (jaystarshot) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1205]: Some tests use compareResuts with verify ordering set to true, so you might need that info

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (jaystarshot) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1267]: > I won't have access to a cteId value so it will stay null or match the name.
Yes same is fine.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (misterjpapa) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1205]: I noticed there isn't any place which calls the compareResults with the checkOrdering field as true yet, which is why I didn't add the boolean checkOrdering field to verifyResults per YAGNI principles. If there are some tests that should be verifying the order, I didn't see them, but I'm okay adding the field for future proofing of the verifyResults method if that is what you want.
Screenshot 2024-07-23 at 8 09 57 AM

Comment (jaystarshot) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1205]: Hmm thats strange, There are a couple of tests which use it https://github.com/prestodb/presto/blob/master/presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java#L677

Comment (jaystarshot) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1205]: Ohh just this one^

Comment (misterjpapa) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1205]: I see it now. Must have missed one. I will make the changes then.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (jaystarshot) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1248]: Need to check the is materialized

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (jaystarshot) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1271]: Nit: Maybe rename to `generateMaterializedCTEInformation`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (misterjpapa) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1248]: The Explain plan doesn't contain information if the CTE is materialized. It is of the form "CTEInfo: [temp: 2 (is_view: false)]". Is there some other way to check?

Comment (jaystarshot) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1248]: Seems like a bug, let me see if we can fix this

Comment (jaystarshot) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1248]: https://github.com/prestodb/presto/pull/23288

Comment (misterjpapa) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1248]: That was quick. I wait for your changes in Issue #23288 to be merged and then add the same materialized check in these tests.

Comment (misterjpapa) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1248]: @jaystarshot - I noticed they merge your commit today. I will update the tests to take into account isMaterialized.

Comment (misterjpapa) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java]: Updated the verifyResults method to check that the expected is_materialized value appears in the verbose Explain plan.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (misterjpapa) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:78]: @jaystarshot - please note that the is_materialized value for the cte named t is false.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (misterjpapa) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:177]: @jaystarshot please note that only the is_materialized value for the cte named "order_platform_data" is false, while the other 4 cte ("urgent", "medium", "high", "low") are all false.

**Comment by jaystarshot** [comment]
Can you please squash all commits into one and clean up the commit message? Commit [standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
In presto we regenerally rebase and merge your commits of the pull request

**Comment by misterjpapa** [comment]
@jaystarshot  - any thoughts why this test would be failing like so (https://github.com/prestodb/presto/actions/runs/10121353652/job/28061341426)? From the output it looks like a call  setSystemProperty(VERBOSE_OPTIMIZER_INFO_ENABLED, "true") against a session "bled" into the session that was used for the generation of the plan so the output is now verbose. 
Screenshot 2024-07-29 at 4 26 33 PM resulting in a failure of comparison. Is that possible?

**Comment by jaystarshot** [comment]
seems to be flaky, i reran and it was successful

**Comment by misterjpapa** [comment]
> Can you please squash all commits into one and clean up the commit message? Commit [standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards). In presto we regenerally rebase and merge your commits of the pull request

I've rebased, squashed all the commits into one pull request, and have updated the commit message.

**Comment by jaystarshot** [comment]
Still seeing 2 commits https://github.com/prestodb/presto/pull/23268/commits.

**Comment by misterjpapa** [comment]
> Still seeing 2 commits https://github.com/prestodb/presto/pull/23268/commits.

I've squashed it to one commit now.

**Comment by jaystarshot** [comment]
https://github.com/prestodb/presto/compare/a774895a44e7a020c2ee0b1cd0f30d56fd53e715..f1014ac79428dad93a528ad6c13288824f1f01fb
Your code regressed to the older version with CteInfoValue.

**Comment by misterjpapa** [comment]
> https://github.com/prestodb/presto/compare/a774895a44e7a020c2ee0b1cd0f30d56fd53e715..f1014ac79428dad93a528ad6c13288824f1f01fb Your code regressed to the older version with CteInfoValue.

I think I've sorted out the squash issues. Apologies for the trouble. I'm not use to dealing with these kinds of git command issues.

**Review by jaystarshot — APPROVED** [review]

**Comment by misterjpapa** [comment]
@jaystarshot  - Thanks for all your efforts here in this PR. Your approval has been over 4 days + 2 days with the weekend ago. Any suggestions on how to get the required code owners to review? Should I make a request in the Dev Slack channel?

**Comment by steveburnett** [comment]
> @jaystarshot - Thanks for all your efforts here in this PR. Your approval has been over 4 days + 2 days with the weekend ago. Any suggestions on how to get the required code owners to review? Should I make a request in the Dev Slack channel?

I suggest yes to your question based on the [Pull Request Checklist](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#pull-request-checklist) in CONTRIBUTING.md: 

_If your pull request does not have a reviewer assigned to it after 4 days, [ask for a review in the #dev channel in Slack](https://join.slack.com/t/prestodb/shared_invite/enQtNTQ3NjU2MTYyNDA2LTYyOTg3MzUyMWE1YTI3Njc5YjgxZjNiYTgxODAzYjI5YWMwYWE0MTZjYWFhNGMwNjczYjI3N2JhM2ExMGJlMWM)._

**Comment by jaystarshot** [comment]
@tdcmeehan Can you please review?

**Review by rschlussel — APPROVED** [review]
approving based on @jaystarshot's review

**Comment by jaystarshot** [comment]
I think we need someone from [team-velox](https://github.com/orgs/prestodb/teams/team-velox) too

File: presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java
Comment (tdcmeehan) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1266]: Do we want to include this in the test?  Won't it increase the log size substantially?

Comment (jaystarshot) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1266]: We need this for confirming that the cte was materialized. This doens't increase logging that much since most of the info is added to the session when this property is on

Comment (tdcmeehan) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestCteExecution.java:1266]: Got it, I had forgotten how this property works.

**Review by tdcmeehan — APPROVED** [review]

**Pull request merged**
